File: dotnet/tests/Buble.Sdk.Tests/GenerationsTests.cs
using Buble.Sdk.Generations;
using Xunit;

namespace Buble.Sdk.Tests;

public sealed class GenerationsTests
{
    [Fact]
    public async Task CreatesFlatGenerationBody()
    {
        var handler = new FakeHttpMessageHandler();
        handler.EnqueueJson("""{"data":{"id":"task_1","status":"pending"}}""");
        using var client = Client(handler);

        await client.Generations.CreateAsync(new CreateGenerationRequest
        {
            Model = "google/nano-banana",
            Mode = "text_to_image",
            Prompt = "hello"
        }.WithParam("aspect_ratio", "1:1").WithParam("output_format", "png"));

        var body = handler.Requests[0].Body;
        Assert.Contains("\"model\":\"google/nano-banana\"", body);
        Assert.Contains("\"aspect_ratio\":\"1:1\"", body);
        Assert.DoesNotContain("\"params\"", body);
    }

    [Fact]
    public void RejectsInternalGenerationFields()
    {
        var request = new CreateGenerationRequest();
        Assert.Throws<UnsupportedGenerationFieldException>(() => request.WithParam("input", new { prompt = "x" }));
    }

    [Fact]
    public async Task WaitsForSuccess()
    {
        var handler = new FakeHttpMessageHandler();
        handler.EnqueueJson("""{"data":{"id":"task_1","status":"processing"}}""");
        handler.EnqueueJson("""{"data":{"id":"task_1","status":"success","result":{"images":[{"url":"https://example.test/out.png"}]}}}""");
        using var client = Client(handler);

        var result = await client.Generations.WaitAsync(
            "task_1",
            new WaitOptions { Interval = TimeSpan.FromMilliseconds(1), Timeout = TimeSpan.FromSeconds(1) });

        Assert.Equal("https://example.test/out.png", result.Data!.Result!.Images![0].Url);
        Assert.Equal("/api/v1/generations/task_1", handler.Requests[0].Uri.AbsolutePath);
    }

    private static BubleClient Client(FakeHttpMessageHandler handler) => new(new BubleClientOptions
    {
        ApiKey = "sk_test",
        BaseUrl = "https://example.test",
        HttpClient = new HttpClient(handler)
    });
}

File: swift/Tests/BubleTests/GenerationsTests.swift
import Buble
import XCTest

final class GenerationsTests: XCTestCase {
    func testCreatesFlatGenerationBody() async throws {
        let transport = try TestHTTPTransport(responses: [
            .json(["data": ["id": "task_1", "status": "pending"]])
        ])
        let client = try makeClient(transport: transport)

        let task = try await client.generations.create(
            try CreateGenerationRequest(model: "google/nano-banana")
                .mode("text_to_image")
                .prompt("A product image")
                .param("aspect_ratio", "1:1")
                .param("output_format", "png")
        )

        XCTAssertEqual(task.data.id, "task_1")
        let requests = await transport.recordedRequests()
        let body = try requestBody(XCTUnwrap(requests.first))
        XCTAssertEqual(body.value(at: ["model"])?.stringValue, "google/nano-banana")
        XCTAssertEqual(body.value(at: ["mode"])?.stringValue, "text_to_image")
        XCTAssertEqual(body.value(at: ["aspect_ratio"])?.stringValue, "1:1")
        XCTAssertNil(body.value(at: ["options"]))
    }

    func testRejectsInternalGenerationFields() {
        XCTAssertThrowsError(try CreateGenerationRequest(model: "google/nano-banana").param("options", ["duration": "5s"])) { error in
            XCTAssertEqual(error as? BubleError, .unsupportedGenerationField(field: "options"))
        }
    }

    func testWaitsUntilSuccess() async throws {
        let transport = try TestHTTPTransport(responses: [
            .json([
                "data": [
                    "id": "task_1",
                    "status": "success",
                    "result": [
                        "images": [
                            ["url": "https://example.com/image.png"]
                        ]
                    ]
                ]
            ])
        ])
        let client = try makeClient(transport: transport)

        let result = try await client.generations.wait("task_1", options: WaitOptions(interval: 0.001, timeout: 1))

        XCTAssertEqual(result.data.result?.images?.first?.url.absoluteString, "https://example.com/image.png")
        let requests = await transport.recordedRequests()
        XCTAssertEqual(requests.first?.url?.path, "/api/v1/generations/task_1")
    }

    func testRaisesOnFailedGeneration() async throws {
        let transport = try TestHTTPTransport(responses: [
            .json([
                "data": [
                    "id": "task_1",
                    "status": "failed",
                    "error": ["message": "provider failed"]
                ]
            ])
        ])
        let client = try makeClient(transport: transport)

        do {
            _ = try await client.generations.wait("task_1", options: WaitOptions(interval: 0.001, timeout: 1))
            XCTFail("Expected failed generation")
        } catch BubleError.generationFailed(let task) {
            XCTAssertEqual(task.error?.message, "provider failed")
        }
    }
}
